server/src/index.js

```diff
@@ -11,52 +11,83 @@ const url = process.env.URL
 app.listen(3000, () => console.log('Listening to port 3000'))
 app.get('/user/:id', (req, res) => res.send(`Welcome to the homepage of user ${req.params.id}.`))
 
+function createPost(req, type) {
+  const post = {}
+  switch (type) {
+    case 0:
+      post.title = req.body.title
+      post.url = req.body.url
+      break
+    case 1:
+      post.title = req.body.title
+      post.body = req.body.body
+      break
+    case 2:
+      post.body = req.body.body
+      post.parent_id = req.body.parentId
+      break
+    default:
+      break
+  }
+  post.type = req.body.type
+  post.timestamp = new Date()
+  post.child_ids = []
+  post.upvote = 0
+  post.downvote = 0
+  return post
+}
+
 const main = async () => {
   const db = await MongoClient.connect(url)
-  // post news and comment
-  app.post('/data', async (req, res) => {
-    const data = {}
-    data.type = req.body.type
-    data.url = req.body.url
-    data.body = req.body.body
-    data.timestamp = new Date()
-    data.upvote = 0
-    data.downvote = 0
-    data.comment_ids = []
-    if (data.type === 'news') {
-      if (!await db.collection('Data').findOne({ url: data.url })) {
-        const r = await db.collection('Data').insertOne(data)
-        console.log('News added to the database.')
-        res.status(201).json({ success: true })
-      } else {
-        console.log('Duplicate news found. Abort.')
-        res.status(409).json({ success: false })
+  // post case 0: news , case 1: top-level comment, case 2: non top-level comment
+  app.post('/posts', async (req, res) => {
+    switch (req.body.type) {
+      case 0: {
+        const duplicate = await db.collection('posts').findOne({ url: req.body.url })
+        if (!duplicate) {
+          const post = createPost(req, req.body.type)
+          const r = await db.collection('posts').insertOne(post)
+          const id = r.insertedId.toHexString()
+          res.status(201).send({ postId: id })
+        } else {
+          res.status(409).send({ postId: duplicate._id.toHexString() })
+        }
+        break
+      }
+      case 1: {
+        const post = createPost(req, req.body.type)
+        const r = await db.collection('posts').insertOne(post)
+        const id = r.insertedId.toHexString()
+        res.status(201).send({ postId: id })
+        break
       }
-    } else {
-      const parentId = ObjectID.createFromHexString(req.body.parent_id, (err, r) => {
-        if (err) {
-          console.log('Parent post not found. Abort.')
-          res.status(409).json({ success: false })
+      case 2: {
+        try {
```

Comment:
You should make `try` block as small as possible. This catches a lot of exceptions, not necessarily from the statement you want. Here it's best to do
```javascript
let parent
try {
  parent = await db.collection('posts').findOne({ _id: parentId })
} ...
```